### PR TITLE
Feature/add conversions

### PR DESCRIPTION
**Pull request by fivetran-jamie**
## PR Overview
**This PR will address the following Issue/Feature:**
no issue made for conversions + https://github.com/fivetran/dbt_facebook_ads/issues/42

**This PR will result in the following new package version:**

v0.8.0

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

## Feature Updates
Introducing...conversion metrics ([PR #43](https://github.com/fivetran/dbt_facebook_ads/pull/43))!
- Adds a `conversion_value` field to each `_report` end model, representing the value of conversions (calculated using the default attribution window set in Meta) that occurred on each day for each ad/campaign/ad set/url/account.
- Creates a `facebook_ads__basic_ad_actions_passthrough_metrics` variable to pass through additional conversion value metrics that are calculated using different attribution windows. 
  - By default, the package includes only the conversion value calculated using the default attribution window, but your report may include calculations using the other windows defined [here](https://developers.facebook.com/docs/marketing-api/reference/ads-action-stats/). See [README](https://github.com/fivetran/dbt_facebook_ads_source/tree/main?tab=readme-ov-file#passing-through-additional-metrics) for details on how to use the new var.

## Documentation
- Documents the ability to transform metrics provided to the `facebook_ads__basic_ad_passthrough_metrics` variable. See [README](https://github.com/fivetran/dbt_facebook_ads/tree/main?tab=readme-ov-file#passing-through-additional-metrics) for details ([PR #43](https://github.com/fivetran/dbt_facebook_ads/pull/43)).

## Under the Hood
- Updated the `quickstart.yml` file to allow for automated Quickstart data model deployments ([PR #40](https://github.com/fivetran/dbt_facebook_ads/pull/40)).
- Updated the PR templates to align with our most up-to-date standards ([PR #43](https://github.com/fivetran/dbt_facebook_ads/pull/43)).

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [ ] dbt run (if incremental models are present)

Before marking this PR as "ready for review" the following have been applied:
- [ ] The appropriate issue has been linked, tagged, and properly assigned
- [x] All necessary documentation and version upgrades have been applied
- [x] docs were regenerated (unless this PR does not include any code or yml updates)
- [x] BuildKite integration tests are passing
- [x] Detailed validation steps have been provided below

### Detailed Validation
Please share any and all of your validation steps:

[Tests](https://github.com/fivetran/dbt_facebook_ads/tree/feature/add-conversions/integration_tests/tests) are passing
image

1. Horizontal integrity test compares total sum of `conversion_value` from each `_report` end model 
2. Consistency tests compare `clicks`, `impressions`, and `spend` across dev and prod in each of the `_report` models

No exceptions needed, just a teeny tiny buffer to account for differences in rounding
image

### CODE REVIEW

File: models/facebook_ads__account_report.sql [hunk 1]
```diff
@@ -30,9 +37,20 @@ joined as (
...
-
         {{ fivetran_utils.persist_pass_through_columns(pass_through_variable='facebook_ads__basic_ad_passthrough_metrics', transform = 'sum') }}
+        , sum(coalesce(conversion_report.conversion_value, 0)) as conversion_value
+
+        {# Adapted from fivetran_utils.persist_pass_through_columns() macro to include coalesces, since the conversion_report is on the right side of a left join #}
+        {% if var('facebook_ads__basic_ad_actions_passthrough_metrics', none) %}
+            {% for field in var('facebook_ads__basic_ad_actions_passthrough_metrics') %}
+                , sum( coalesce({{ field.alias if field.alias else field.name }}, 0)) as {{ field.alias if field.alias else field.name }}
+            {% endfor %}
+        {% endif %}
```
Comment (fivetran-joemarkiewicz) [inline models/facebook_ads__account_report.sql:48]: What are your thoughts on consolidating this into a macro?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,7 +1,17 @@
-# dbt_facebook_ads v0.NEXT.RELEASE
+# dbt_facebook_ads v0.8.0
+
+## Feature Updates
+Introducing...conversion metrics ([PR #43](https://github.com/fivetran/dbt_facebook_ads/pull/43))!
+- Adds a `conversion_value` field to each `_report` end model, representing the value of conversions (calculated using the default attribution window set in Meta) that occurred on each day for each ad/campaign/ad set/url/account.
+- Creates a `facebook_ads__basic_ad_actions_passthrough_metrics` variable to pass through additional conversion value metrics that are calculated using different attribution windows. 
+  - By default, the package includes only the conversion value calculated using the default attribution window, but your report may include calculations using the other windows defined [here](https://developers.facebook.com/docs/marketing-api/reference/ads-action-stats/). See [README](https://github.com/fivetran/dbt_facebook_ads_source/tree/main?tab=readme-ov-file#passing-through-additional-metrics) for details on how to use the new var.
```
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:7]: Suggestion to not truncate variable
```suggestion
  - By default, the package includes only the conversion value calculated using the default attribution window, but your report may include calculations using the other windows defined [here](https://developers.facebook.com/docs/marketing-api/reference/ads-action-stats/). See [README](https://github.com/fivetran/dbt_facebook_ads_source/tree/main?tab=readme-ov-file#passing-through-additional-metrics) for details on how to use the new variable.
```

File: README.md [hunk 3]
```diff
@@ -31,7 +31,8 @@ The following table provides a detailed list of all models materialized within t
...
 | [facebook_ads__ad_set_report](https://fivetran.github.io/dbt_facebook_ads/#!/model/model.facebook_ads.facebook_ads__ad_set_report)            | Each record in this table represents the daily performance at the ad set level. |
 | [facebook_ads__ad_report](https://fivetran.github.io/dbt_facebook_ads/#!/model/model.facebook_ads.facebook_ads__ad_report)            | Each record in this table represents the daily performance at the ad level. |
-| [facebook_ads__utm_report](https://fivetran.github.io/dbt_facebook_ads/#!/model/model.facebook_ads.facebook_ads__utm_report)            | Each record in this table represents the daily performance of URLs at the ad level. |
+| [facebook_ads__url_report](https://fivetran.github.io/dbt_facebook_ads/#!/model/model.facebook_ads.facebook_ads__url_report)            | Each record in this table represents the daily performance of URLs at the ad level. |
+| [facebook_ads__url_tags](https://fivetran.github.io/dbt_facebook_ads/#!/model/model.facebook_ads.facebook_ads__utm_report)            | Each record in this table is a unique combination of creative_id and the corresponding key, value, and type values contained in the url_tags field. |
```
Comment (fivetran-joemarkiewicz) [inline README.md:35]: Great catch!

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,7 +1,17 @@
...
+
+## Documentation
+- Documents the ability to transform metrics provided to the `facebook_ads__basic_ad_passthrough_metrics` variable. See [README](https://github.com/fivetran/dbt_facebook_ads/tree/main?tab=readme-ov-file#passing-through-additional-metrics) for details ([PR #43](https://github.com/fivetran/dbt_facebook_ads/pull/43)).
 
 ## Under the Hood
-- Updated the `quickstart.yml` file to allow for automated Quickstart data model deployments.  ([PR #40](https://github.com/fivetran/dbt_facebook_ads/pull/40))
+- Updated the `quickstart.yml` file to allow for automated Quickstart data model deployments ([PR #40](https://github.com/fivetran/dbt_facebook_ads/pull/40)).
+- Updated the PR templates to align with our most up-to-date standards ([PR #43](https://github.com/fivetran/dbt_facebook_ads/pull/43)).
```
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:14]: Not necessary for this initial review, but let's include a Contributors section here and make sure we give Seer a shout out. Will align more on what we should say once we chat with Seer.

**Review by fivetran-joemarkiewicz — APPROVED** [review]
@fivetran-jamie these changes look great! I just have a few minor suggestions, but none related to the actual code. I see this being a huge value add for our users!

Once you address the comments in this review let's hold off merging until the other updates in the Google and LinkedIn packages are ready as well. Thanks and great job!

**Comment by fivetran-catfritz** [comment]
@fivetran-jamie @fivetran-joemarkiewicz Now might be a good time to delete the defunct [2nd reviewer bot](https://github.com/fivetran/dbt_facebook_ads/blob/main/.github/workflows/add-2nd-reviewer.yml).

**Comment by fivetran-joemarkiewicz** [comment]
> @fivetran-jamie @fivetran-joemarkiewicz Now might be a good time to delete the defunct [2nd reviewer bot](https://github.com/fivetran/dbt_facebook_ads/blob/main/.github/workflows/add-2nd-reviewer.yml).

Agreed! @fivetran-jamie would you be able to remove that from this branch.

File: integration_tests/dbt_project.yml [hunk 5]
```diff
@@ -9,8 +9,18 @@ vars:
...
   facebook_ads_creative_history_identifier: "facebook_ads_creative_history_data"
+  facebook_ads__basic_ad_actions_passthrough_metrics: 
+    - name: 'inline'
+      alias: 'conversion_value_inline'
+    - name: '_1_d_view'
+    - name: '_7_d_click'
+    - name: '_28_d_click'
+  facebook_ads__conversion_action_types:
+    - name: lead
+    - pattern: offsite_conversion%
```
[the remark this answers is not in the document]
Comment (fivetran-jamie) [reply integration_tests/dbt_project.yml:23]: ah yeah i only added them locally

File: models/facebook_ads__account_report.sql (hunk 1 above)
Comment (fivetran-jamie) [reply models/facebook_ads__account_report.sql:48]: yeah this exact code is used in like 5 models so i think a macro is very appropriate!

File: models/facebook_ads__account_report.sql [hunk 6]
```diff
@@ -30,9 +37,20 @@ joined as (
         sum(report.clicks) as clicks,
         sum(report.impressions) as impressions,
         sum(report.spend) as spend
-
         {{ fivetran_utils.persist_pass_through_columns(pass_through_variable='facebook_ads__basic_ad_passthrough_metrics', transform = 'sum') }}
+        , sum(coalesce(conversion_report.conversion_value, 0)) as conversion_value
```
[the remark this answers is not in the document]
Comment (fivetran-jamie) [reply models/facebook_ads__account_report.sql:41]: yeah, though i have actually yet to see a `null` (or even `0`) conversion value recorded in `basic_ad_actions`

the individual/passthrough values are null when representing 0 however
image

File: CHANGELOG.md (hunk 4 above)
Comment (fivetran-jamie) [reply CHANGELOG.md:14]: added a section

File: models/facebook_ads__account_report.sql (hunk 1 above)
Comment (fivetran-jamie) [reply models/facebook_ads__account_report.sql:48]: added

File: models/facebook_ads__account_report.sql (hunk 6 above)
[the remark this answers is not in the document]
Comment (fivetran-joemarkiewicz) [reply models/facebook_ads__account_report.sql:41]: Thanks for confirming. Good to see we haven't seen `null` in the conversion value, but great to see we are being proactive with this logic. This looks good to me!

**Comment by maxtuzz** [comment]
Great effort! Looking forward to this release.

**Comment by fivetran-jamie** [comment]
FYI for anyone subscribed here, this is currently blocked. 

We are working with Engineering to ensure that the conversion monetary values (`action_values`) are included in a default report and therefore accessible to the package (currently, only the raw # of conversions are supported by the `basic_ad_actions` report).

File: docs/run_results.json
Comment (fivetran-joemarkiewicz) [inline docs/run_results.json:1]: Let's remove this file

File: integration_tests/tests/consistency/consistency_account_report.sql [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    tags="fivetran_validations",
+    enabled=var('fivetran_validation_tests_enabled', false)
+) }}
+
+with prod as (
+    select
+        account_id,
+        sum(coalesce(clicks, 0)) as clicks, 
+        sum(coalesce(impressions, 0)) as impressions,
+        sum(coalesce(spend, 0)) as spend
```
Comment (fivetran-joemarkiewicz) [inline integration_tests/tests/consistency/consistency_account_report.sql:11]: Request to add conversions in but comment them out for this release so we may use them in future updates.

Comment (fivetran-joemarkiewicz) [reply integration_tests/tests/consistency/consistency_account_report.sql:11]: Same comment for other validation tests

File: packages.yml [hunk 8]
```diff
@@ -1,3 +1,8 @@
 packages:
-- package: fivetran/facebook_ads_source
-  version: [">=0.7.0", "<0.8.0"]
+# - package: fivetran/facebook_ads_source
+#   version: [">=0.8.0", "<0.9.0"]
+# - local: ../Facebook_ads/dbt_facebook_ads_source
+# - local: ../../dbt_facebook_ads_source
+- git: https://github.com/fivetran/dbt_facebook_ads_source.git
+  revision: feature/add-conversions
+  warn-unpinned: false
```
Comment (fivetran-joemarkiewicz) [inline packages.yml:8]: Reminder to clean and swap before merge

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-jamie great work on this PR and you're nearly there to releasing full support for conversions in our ad reporting packages!!

I have a few comments and suggestions in line. Additionally, I noticed I ran into a similar error from before where I run into an issue when trying to passthrough a default conversion value from this PR into the `facebook_ads__basic_ad_passthrough_metrics` variable. Would you be able to look into this and confirm if this is intentional or if there are some adjustments we need to make to this logic within the models. 

image

Thanks!

**Comment by fivetran-jamie** [comment]
> I noticed I ran into a similar error from before where I run into an issue when trying to passthrough a default conversion value from this PR into the facebook_ads__basic_ad_passthrough_metrics variable. Would you be able to look into this and confirm if this is intentional or if there are some adjustments we need to make to this logic within the models.

@fivetran-joemarkiewicz Yeah this is the expected behavior, as conversions are explicitly not included in `basic_ad` according to our [docs](https://fivetran.com/docs/connectors/applications/facebook-ads/prebuilt-reports). However, I suppose there is nothing stopping someone from adding `conversions` as a passthrough field, even if null (especially if they were using passthrough columns to add conversions to the ad_reporting models). 

To avoid duplicate column errors, we can employ the same `_c` suffix method as [ad_reporting](https://github.com/fivetran/dbt_ad_reporting/pull/122/), or we can just choose to select our version of `conversions` since any previous field was likely null

**Comment by fivetran-joemarkiewicz** [comment]
> conversions are explicitly not included in basic_ad according to our [docs](https://fivetran.com/docs/connectors/applications/facebook-ads/prebuilt-reports).

Thanks for calling this out @fivetran-jamie! I was going through and stress testing so I see how I tested something that realistically should never occur. I would recommend if the conversion fields should not exist in the BASIC_AD table, then I feel comfortable giving our default value the preference and not including the `_c` field name as it will most likely be null. 

Thanks for clarifying this!

File: integration_tests/tests/consistency/consistency_account_report.sql (hunk 7 above)
Comment (fivetran-jamie) [reply integration_tests/tests/consistency/consistency_account_report.sql:11]: added

**Review by fivetran-joemarkiewicz — APPROVED** [review]
LGTM!

File: README.md [hunk 9]
```diff
@@ -89,19 +93,62 @@ vars:
...
 
 #### Passing Through Additional Metrics
-By default, this package will select `clicks`, `impressions`, and `cost` from the source reporting tables to store into the staging models. If you would like to pass through additional metrics to the staging models, add the below configurations to your `dbt_project.yml` file. These variables allow for the pass-through fields to be aliased (`alias`) if desired, but not required. Use the below format for declaring the respective pass-through variables:
+By default, this package will select `clicks`, `impressions`, `cost`, and conversion `value` (using the default attribution window) from the source reporting tables (`BASIC_AD`, `BASIC_AD_ACTIONS`, and `BASIC_AD_ACTION_VALUES`) to store into the output models. If you would like to pass through additional metrics to the output models, add the below configurations to your `dbt_project.yml` file. These variables allow for the pass-through fields to be aliased (`alias`) and transformed (`transform_sql`) if desired, but not required. Only the `name` of each metric field is required. Use the below format for declaring the respective pass-through variables:
```
Comment (fivetran-reneeli) [inline README.md:96]:
```suggestion
By default, this package will select `clicks`, `impressions`, `cost`, `conversion`, and conversion `value` (using the default attribution window) from the source reporting tables (`BASIC_AD`, `BASIC_AD_ACTIONS`, and `BASIC_AD_ACTION_VALUES`) to store into the output models. If you would like to pass through additional metrics to the output models, add the below configurations to your `dbt_project.yml` file. These variables allow for the pass-through fields to be aliased (`alias`) and transformed (`transform_sql`) if desired, but not required. Only the `name` of each metric field is required. Use the below format for declaring the respective pass-through variables:
```

File: models/facebook_ads__campaign_report.sql [hunk 10]
```diff
@@ -48,10 +55,18 @@ joined as (
...
+        sum(coalesce(conversion_report.conversions_value, 0)) as conversions_value
+
+        {{ facebook_ads_persist_pass_through_columns(pass_through_variable='facebook_ads__basic_ad_passthrough_metrics', transform = 'sum', exclude_fields=['conversions', 'conversions_value']) }}
+        {{ facebook_ads_persist_pass_through_columns(pass_through_variable='facebook_ads__basic_ad_actions_passthrough_metrics', transform = 'sum', coalesce_with=0) }}
+        {{ facebook_ads_persist_pass_through_columns(pass_through_variable='facebook_ads__basic_ad_action_values_passthrough_metrics', transform = 'sum', coalesce_with=0) }}
 
-        {{ fivetran_utils.persist_pass_through_columns(pass_through_variable='facebook_ads__basic_ad_passthrough_metrics', transform = 'sum') }}
     from report 
+    left join conversion_report
+        on report.date_day = conversion_report.date_day
```
Comment (fivetran-reneeli) [inline models/facebook_ads__campaign_report.sql:68]: do we need to join by source_relation here as well?

File: models/facebook_ads__url_report.sql [hunk 11]
```diff
@@ -71,10 +78,18 @@ joined as (
...
+        sum(coalesce(conversion_report.conversions_value, 0)) as conversions_value
+
+        {{ facebook_ads_persist_pass_through_columns(pass_through_variable='facebook_ads__basic_ad_passthrough_metrics', transform = 'sum', exclude_fields=['conversions', 'conversions_value']) }}
+        {{ facebook_ads_persist_pass_through_columns(pass_through_variable='facebook_ads__basic_ad_actions_passthrough_metrics', transform = 'sum', coalesce_with=0) }}
+        {{ facebook_ads_persist_pass_through_columns(pass_through_variable='facebook_ads__basic_ad_action_values_passthrough_metrics', transform = 'sum', coalesce_with=0) }}
 
-        {{ fivetran_utils.persist_pass_through_columns(pass_through_variable='facebook_ads__basic_ad_passthrough_metrics', transform = 'sum') }}
     from report
+    left join conversion_report
+        on report.date_day = conversion_report.date_day
```
Comment (fivetran-reneeli) [inline models/facebook_ads__url_report.sql:91]: source_relation join?

**Review by fivetran-reneeli — COMMENTED** [review]
Just had some questions!

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,7 +1,51 @@
...
+```
+
+>**Note**: If you were previously utilizing the `facebook_ads__basic_ad_passthrough_metrics` variable to include a (likely null) field called `conversions` or `conversions_value`, the package's version of these fields will take precedence over yours. To continue including your old conversions fields, change the `alias` of the passthrough field to another name.
+
+- Creates the new `facebook_ads__basic_ad_actions_passthrough_metrics` and `facebook_ads__basic_ad_action_values_passthrough_metrics` variables to pass through additional conversion value metrics that are calculated using different attribution windows. 
+  - By default, the package includes only the conversion values calculated using the default attribution window set in Meta, but your report may include calculations using the other windows defined [here](https://developers.facebook.com/docs/marketing-api/reference/ads-action-stats/). See [README](https://github.com/fivetran/dbt_facebook_ads/tree/main?tab=readme-ov-file#passing-through-additional-metrics) for details on how to use the new variables.
```
[the remark this answers is not in the document]
Comment (fivetran-jamie) [reply CHANGELOG.md:33]: adding link

File: models/facebook_ads__url_report.sql (hunk 11 above)
Comment (fivetran-jamie) [reply models/facebook_ads__url_report.sql:91]: doh!

File: models/facebook_ads__campaign_report.sql (hunk 10 above)
Comment (fivetran-jamie) [reply models/facebook_ads__campaign_report.sql:68]: jeez I can't believe I forgot these, good catch

**Review by fivetran-reneeli — APPROVED** [review]
LGTM

**Pull request merged**